### DISCUSSION TITLE
Running 3rd-party applications on MacOS and Filesystem blocking on MacOS

### DISCUSSION THREAD

**Discussion by thonglepas** [SafeExamBrowser/seb-mac#418] [Q&A]
Hi,

We are planning to create an assessment using SEB. The assessment will be run on candidate's BYO devices. During the assessment, the candidate will require to use some 3rd-party applications (Adobe apps, etc.) to complete the assessment. My questions are:

- Is there any way we can allow 3rd-party applications to run in SEB in MacOS?
- Due to the ability to access filesystem via 3rd-party applications, is there any ways we can restrict the access for the filesystem except for some certain folders in SEB?

Thank you very much!

**Comment by danschlet** [comment]
- Running 3rd-party applications on macOS is possible as of SEB 3.4, see the current beta version on the [Releases page](https://github.com/SafeExamBrowser/seb-mac/releases).
- SEB 3.4 can prevent the system-provided Open File/Save As dialogs to be displayed. It also resets the open documents feature of most macOS applications (so if you had documents open when you used an app, those won't open when you start that app in SEB). 
- Some apps should only be started in SEB by downloading and opening a document of that app, as otherwise they will show a "new file" window which lists recent files the user worked on. This is for example the case with Microsoft Office apps. There is a setting in SEB's Settings/Applications/Permitted Processes. For such apps, you should deselect "Allow manual start".
- SEB 3.4 currently doesn't show a custom folder which would be accessible in this Multi App Mode. But it has the feature that you can only select files for upload in a web assessment solution with the "Choose file" for upload button which were previously downloaded inside SEB. So if working with third party applications, you should provide download links for template files you want students to work on with 3rd-party apps. When they click such links, these files get downloaded and open in the according 3rd-party app. Then they have to save the file they worked on (with cmd-S) with the same filename and can upload that file to the web assessment solution.

**Reply by soutienti** [reply]
Thank you very much, it works really well and on iPad too! my users with special needs can use the software they need.

**Comment by JonasPut** [comment]
Hi
Which settings should I allow when permitting apps like Desmos or Bookwidget on SEB?
When I configure this on Macbook  I can see a toolbar with the apps Desmos and Bookwidget but they do not open when clicked.
When i configure this on an Ipad I do not see the apps on the toolbar. What am I doing wrong?

Thx for helping.

**Reply by danschlet** [reply]
We have to add documentation for these new features. In general we cannot guarantee if macOS apps will actually work with this feature. Regarding settings: Just use the "Choose Application..." button in SEB macOS Settings / Applications / Permitted Processes and SEB will populate the Identifier and Executable fields and will suggest you to change other settings, so the AAC Multi-App mode should work correctly. You must use the SEB macOS settings currently to set up all options correctly (don't use SEB Windows Config Tool).

I successfully tested Excel, Numbers, Pages and some other apps on Mac. One or two I tried didn't work. I don't think there is a lot I can do if some don't work. Maybe allowing some additional bundle IDs in case the app uses several processes might help, but this has to be tested in a trial-and-error manner.

<img width="762" alt="Screenshot 2025-01-06 at 15 32 21" src="https://github.com/user-attachments/assets/974a4c84-d114-43cb-9118-7012080d5767" />

In iPadOS, you just go to the home screen (home button or swipe up gesture) and you will see only SEB and the additional apps. This is how the AAC Multi-App mode is implemented in iPadOS, we probably won't be able to implement it exactly how it looks on macOS (or Windows) with icons in the SEB Dock/taskbar. Note that if you enter a non-existing bundle ID in Identifier, on iPadOS the AAC Assessment Mode won't start correctly (there will be an error message after a some seconds). Apple will have to improve this, SEB cannot check if the app for a particular bundle ID is actually installed on the device.

**Reply by danschlet** [reply]
PS: Likely proper, native macOS apps following Apple's Human Interface Guidelines and other best practices will work with AAC Multi-App mode. Apps using cross-platform nonsense frameworks like Elektron or software like Adobe's apps which use a dozen of background processes and outdated technologies like Java applets won't work.

**Reply by MarkBlondal** [reply]
Sorry to hijack the thread, but the question is related to the OP's question.

@danschlet If I can get a Java application (TI-Nspire CX CAS Student Software) to run, but aren't able to trigger the "Save as" window.
Is that due to the limitations of SEB and Java applications, or are there specific settings that need to be done for that part to work?

**Reply by danschlet** [reply]
You have to enable the setting "Allow open/close panel" (see the screen shot above). In case it still doesn't work, then it might be an incompatibility with the AAC Assessment Mode. On older macOS versions (like 11 Big Sur) I observed that some windows of or triggered by additional apps were hidden by the system.

**Reply by MarkBlondal** [reply]
It's running the same both on Monterey and Tahoe (Tried with two different machines.

I can trigger the save function, but the windows where I accept the save command doesn't appear until after I close the SEB session. So it seems like it's somehow running in the background and just not appearing.

<img width="1487" height="939" alt="Skærmbillede 2025-09-23 kl  07 24 27" src="https://github.com/user-attachments/assets/5e764635-5792-4e0e-a637-4f4e891710af" />

**Reply by danschlet** [reply]
Maybe there is another process involved, which is displaying the save dialog. That process would also need to be added to "Permitted Processes", maybe you can disable "Icon in taskbar" for that helper process. But no idea if this would work and which process that would be. You can use Activity Monitor and search for processes with the name of that app or something including "java". Java applications are a mess, on Windows we were successful making some run in the past with such custom settings in Permitted Processes. On macOS I didn't try any Java-based applications yet. The AAC Assessment Mode Multi-App feature will most likely work with proper macOS AppKit applications, if apps created with some of these cross-platform technologies will work is doubtful.

Do you have to use "Save As"? Maybe at least "Save" works (I usually would try the cmd-S shortcut)? We would place template files into an LMS/web assessment system and let students download these and then save their modified file with "Save". Then no file open/save as dialog is needed, so most apps wouldn't give students access to the file system.

**Reply by MarkBlondal** [reply]
Hi Danschlet

The reason for the separate post, was to not hijack the original post since my issue deviated somewhat.
And I wanted to append some files, to facilitate some debugging, since that usually helps in locating the root cause.

No matter if I use the Floppy icon, or the shortcut cmd+S, the window refuses to appear until I close the SEB session.

I tried getting some AI tools to browse through the log file, and it just gives this response:
Okay, that is very strange.
The log file confirms that you have removed the manual permission, so now only the general "Allow open/save panel" function should handle it. Yet, we still see exactly the same problem in the log:
2025/10/21 14:29:53:196 Started application with bundle ID: com.apple.appkit.xpc.openAndSavePanelService
2025/10/21 14:29:53:196 Successfully terminated app ... executable URL: .../com.apple.appkit.xpc.openAndSavePanelService.xpc/
This strongly suggests that the built-in "Allow open/save panel" function is simply not working correctly in this specific situation – either due to your SEB version (3.6 Build 156D0), your macOS version (26.0.1 Build 25A362 in the log), or the way Nspire calls the dialog.

We have exhausted the logical configuration options for manually allowing the process, and the automatic function isn't working either.

Is the openAndSavePanelService not the standard service responsible for saving files when using SEB, or does it have another way it handles saving files?

**Reply by danschlet** [reply]
You were right creating a separate issue when you actually find a bug. I initially didn't read the description properly, my apologies. I added some more information to https://github.com/SafeExamBrowser/seb-mac/issues/550.
